src/core/qgslegendrenderer.cpp

```diff
@@ -494,36 +499,53 @@ int QgsLegendRenderer::setColumns( QList<LegendComponentGroup> &componentGroups
   double averageSpaceAboveGroups = 0;
   if ( componentGroups.size() > targetNumberColumns )
     averageSpaceAboveGroups = totalSpaceAboveGroups / ( componentGroups.size() );
-  // Correct the totalHeight using the number of columns because the first item
-  // in each column does not get any space above it
-  totalHeight -= targetNumberColumns * averageSpaceAboveGroups;
 
+  double totalRemainingGroupHeight = totalGroupHeight;
+  double totalRemainingSpaceAboveGroups = totalSpaceAboveGroups;
   for ( int i = 0; i < componentGroups.size(); i++ )
   {
-    LegendComponentGroup group = componentGroups.at( i );
-    double currentHeight = currentColumnHeight;
+    const LegendComponentGroup &group = componentGroups.at( i );
+    const double currentGroupHeight = group.size.height();
+    const double spaceAboveCurrentGroup = spaceAboveGroup( group );
+
+    totalRemainingGroupHeight -= currentGroupHeight;
+    totalRemainingSpaceAboveGroups -= spaceAboveCurrentGroup;
+
+    double currentColumnHeightIfGroupIsIncluded = currentColumnHeight;
     if ( currentColumnGroupCount > 0 )
-      currentHeight += spaceAboveGroup( group );
-    currentHeight += group.size.height();
+      currentColumnHeightIfGroupIsIncluded += spaceAboveCurrentGroup;
+    currentColumnHeightIfGroupIsIncluded += currentGroupHeight;
+
+    const int numberRemainingGroupsIncludingThisOne = componentGroups.size() - i;
+    const int numberRemainingColumnsIncludingThisOne = numberAutoPlacedBreaks + 1 - autoPlacedBreaks;
+    const int numberRemainingColumnBreaks = numberRemainingColumnsIncludingThisOne - 1;
 
-    int numberRemainingGroups = componentGroups.size() - i;
+    const double averageRemainingSpaceAboveGroups = numberRemainingGroupsIncludingThisOne > 1 ? ( totalRemainingSpaceAboveGroups / ( numberRemainingGroupsIncludingThisOne - 1 ) ) : 0;
+    const double estimatedRemainingSpaceAboveGroupsWhichWontBeUsedBecauseGroupsAreFirstInColumn = numberRemainingColumnBreaks * averageRemainingSpaceAboveGroups;
+    const double estimatedRemainingTotalHeightAfterThisGroup = totalRemainingGroupHeight
+        + totalRemainingSpaceAboveGroups
+        - estimatedRemainingSpaceAboveGroupsWhichWontBeUsedBecauseGroupsAreFirstInColumn;
```

Comment:
@nyalldawson , isThisTheLongestVariableNameWeHaveInOurSourceTree? :) Joking apart, nice job making the code more readable via variable renaming and use of const.

Reply:
:upside_down_face: in all honesty, I couldn't think of a more concise way to name this would wouldn't result in me having to read this code 100x over again to work out exactly what's happening here!!

Reply:
Yeah, sometimes you just have to settle for a 78 characters variable name :smile: